### ISSUE TITLE
Provide cluster endpoint addresses with instaclustr_cluster resource

### ISSUE THREAD

**Issue by ghost** [instaclustr/terraform-provider-instaclustr#50]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

We would like to use generated IP addresses of the kafka cluster elsewhere (e.g. save them in the secrets manager as values for applications to retrieve or use them in Glue AWS resource).

Provide an attribute from the cluster resource with these values. Looks like these are random in the subnets from the given VPC CIDR and we cannot at least set them statically.

Eventually please provide a data_source for the cluster to retrieve the endpoints.

### New or Affected Resource(s)

* instaclustr_cluster

### Potential Terraform Configuration

```hcl
resource "aws_glue_connection" "kafka" {
  connection_type = "KAFKA"
  connection_properties = {
    KAFKA_BOOTSTRAP_SERVERS = instaclustr_cluster.kafka.endpoints <---- this is one of our use cases
    KAFKA_SSL_ENABLED       = "false"
  }
  name = "${var.name}-kafka"
  physical_connection_requirements {
    subnet_id = var.private_subnets[0]
    security_group_id_list = [aws_security_group.glue_kafka.id]
    availability_zone = var.azs[0] # this is useless but required (inherited from subnet)
  }
}
```

### References

* #0000

**Comment by ProfMoo** [comment]
Another great use case for this would be generating DNS records based on the IPs.

**Comment by jurgenweber** [comment]
another would be creating a kubernetes configmap or secret wit the values for deployments/pods to pick up.

**Comment by jurgenweber** [comment]
after a bit of digging, I can see it is partially in here:

https://github.com/instaclustr/terraform-provider-instaclustr/blob/9d114e03610650cb9392f1e4005eabfa254b4000/instaclustr/resource_cluster.go#L386-L393

but it seems to be only pulling the IP address of the first node.

Ok, I took a stab at this today but I have a problem, it always seems to be trying to modify/update the current cluster even though it should not (as far as I can tell):

https://github.com/instaclustr/terraform-provider-instaclustr/compare/master...jurgenweber:master

**Comment by peter-toms** [comment]
thanks @janlud and @jurgenweber if I understand correctly you are interested in seeing the bootstrap servers as an output value from the instaclustr_cluster ?

**Comment by jurgenweber** [comment]
yeah ,I am nearly done....

https://github.com/instaclustr/terraform-provider-instaclustr/compare/master...jurgenweber:master

I just have one bug I am trying to work out but I am unsure to where I am gong wrong.

We have 'private_network_cluster = true'.... but I am always seeing a change for:

```
        private_network_cluster   = true
      + public_contact_addresses  = (known after apply)
        rack_allocation           = {
            "nodes_per_rack"  = "1"
            "number_of_racks" = "3"
        }
```
any ideas?

**Comment by peter-toms** [comment]
Wow thanks @jurgenweber we really appreciate it!  Ive added a comment on your PR (I've had a similar issue previously.  It thinks you are trying to add an optional input value to the schema and because its always changing that will pop the change constantly).  Cheers.

**Comment by jurgenweber** [comment]
cool, but I see no comments in the PR! 👍

**Comment by peter-toms** [comment]
helps if it submitted the review... try again now

**Comment by peter-toms** [comment]
adding internal reference INS-8760

**Comment by ProfMoo** [comment]
@peter-toms Any update here?

**Comment by peter-toms** [comment]
Hi @ProfMoo, we have this one in our backlog, unfortunately I don't have any specific timeframes I can share at this time.  Will keep you updated as we progress.

**Comment by jan-ludvik-deltatre** [comment]
Hello, any progress? I think this is fundamental to the provider - I created a cluster and now I need to pass the broker addresses elsewhere. Seems to me like this should have higher priority.

**Comment by peter-toms** [comment]
Hi guys, thanks for the message and your patience on this one!  We have prioritised this ticket as part of our current sprint backlog and expect to deliver it within the next fortnight.  At the moment we are anticipating to implement this with output parameters which will enable you to gather these addresses for downstream use.

Thanks, will keep you updated on our progress.

**Comment by jurgen-weber-deltatre** [comment]
@peter-toms How did you go? How long are your sprints? :)

**Comment by peter-toms** [comment]
Hi @jurgen-weber-deltatre, thanks and apologies for the delay in updating here.  We are in the final stages of development now and are expecting early next week to have the feature ready.  Will keep you updated on our progress.  Cheers.

**Comment by Chrismelba-Instaclustr** [comment]
Hi there. This should be fixed in release 1.9.6! Thanks so much for bringing it to our attention.

**Comment by Chrismelba-Instaclustr** [comment]
Just a quick note. For this to work as intended you'll need to add "wait_for_state="RUNNING"" to the clusters you wish to get the contact points for, as these IPs aren't known until the cluster is running

**Comment by jurgen-weber-deltatre** [comment]
ok, but what is the attribute?

**Comment by Chrismelba-Instaclustr** [comment]
Sorry, the attributes of clusters public_contact_point and private_contact_point should now populate with one IP per AZ. You can use these as output variables in your terraform state.

**Comment by jurgen-weber-deltatre** [comment]
Thanks, right now I am getting this funky error:

> Error: missing expected [

got me blown

**Comment by jurgen-weber-deltatre** [comment]
ok, so I worked out what this error is, as per: https://github.com/alexkappa/terraform-provider-auth0/issues/157#issuecomment-603440102

So in your PR; https://github.com/instaclustr/terraform-provider-instaclustr/commit/6cead3bdb3268570281718c2aa6489bc3e6516f8#diff-9c16cc48a81c7cf880778bd3ca138feecbdfebee18764cdcc5a1c91c81fbe444R92-R101

you change the type of a field, so if you have an existing state with that attribute it will error. I had to manually change my state and convert that field from a string to a [].

Anyone updating to this version of the provider is going to have the problem, I think. Not sure as I forked and maintained my own to have the output attribute but that was a while ago now. Someone else should confirm this.

That said, the lesson here is don't change the types of existing attributes. Just make a new attribute with the new thing/type. :)
